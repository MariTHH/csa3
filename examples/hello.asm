.data:
    str hw: 'H', 'e', 'l', 'l', 'o', ' ', 'w', 'o', 'r', 'l', 'd', '!'
    num cur_ind: 0

.text:
      mov [hw] cur_ind
      add cur_ind 0
loop: out# cur_ind
      cmp* cur_ind 0
      je end
      jmp loop
end:  hlt



