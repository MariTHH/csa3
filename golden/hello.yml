source: |-
  .data:
      str hw: 'H', 'e', 'l', 'l', 'o', ' ', 'w', 'o', 'r', 'l', 'd', '!'
      num cur_ind: 0

  .text:
      mov [hw] cur_ind
      add cur_ind 0
  loop: out# cur_ind
      cmp* cur_ind 0
      je end
      jmp loop
  end:  hlt


input: |

code: |-
  [
      {
          "opcode": "mov",
          "addr": 0,
          "args": [
              14,
              13
          ]
      },
      {
          "opcode": "add",
          "addr": 1,
          "args": [
              13,
              15
          ]
      },
      {
          "opcode": "out_rel",
          "addr": 2,
          "args": [
              13
          ]
      },
      {
          "opcode": "cmp_rel_inc",
          "addr": 3,
          "args": [
              13,
              16
          ]
      },
      {
          "opcode": "je",
          "addr": 4,
          "args": [
              6
          ]
      },
      {
          "opcode": "jmp",
          "addr": 5,
          "args": [
              2
          ]
      },
      {
          "opcode": "hlt",
          "addr": 6,
          "args": []
      }
  ]
output: |
  source LoC: 38 code instr: 7
  ============================================================
  Hello world!
  instr_counter: 53 ticks: 162
log: |
  DEBUG    root:machine.py:277 {TICK: 0, ADDR: 0, PC: 0, OUT: 72, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 1, ADDR: 14, PC: 0, OUT: 0, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 2, ADDR: 13, PC: 0, OUT: 0, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 3, ADDR: 13, PC: 1, OUT: 0, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 4, ADDR: 13, PC: 1, OUT: 0, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 5, ADDR: 15, PC: 1, OUT: 0, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 6, ADDR: 13, PC: 1, OUT: 0, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 7, ADDR: 13, PC: 2, OUT: 0, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 8, ADDR: 13, PC: 2, OUT: 0, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 9, ADDR: 0, PC: 2, OUT: 72, ACC: 72} 
  DEBUG    root:machine.py:100 output: '' << 'H'
  DEBUG    root:machine.py:138 {TICK: 10, ADDR: 0, PC: 2, OUT: 72, ACC: 72} 
  DEBUG    root:machine.py:138 {TICK: 11, ADDR: 0, PC: 3, OUT: 72, ACC: 72} 
  DEBUG    root:machine.py:138 {TICK: 12, ADDR: 13, PC: 3, OUT: 0, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 13, ADDR: 0, PC: 3, OUT: 72, ACC: 72} 
  DEBUG    root:machine.py:138 {TICK: 14, ADDR: 16, PC: 3, OUT: 0, ACC: -72} 
  DEBUG    root:machine.py:138 {TICK: 15, ADDR: 13, PC: 3, OUT: 0, ACC: -72} 
  DEBUG    root:machine.py:138 {TICK: 16, ADDR: 13, PC: 3, OUT: 1, ACC: -72} 
  DEBUG    root:machine.py:138 {TICK: 17, ADDR: 13, PC: 4, OUT: 1, ACC: -72} 
  DEBUG    root:machine.py:138 {TICK: 18, ADDR: 13, PC: 5, OUT: 1, ACC: -72} 
  DEBUG    root:machine.py:138 {TICK: 19, ADDR: 13, PC: 2, OUT: 1, ACC: -72} 
  DEBUG    root:machine.py:138 {TICK: 20, ADDR: 13, PC: 2, OUT: 1, ACC: 1} 
  DEBUG    root:machine.py:138 {TICK: 21, ADDR: 1, PC: 2, OUT: 101, ACC: 101} 
  DEBUG    root:machine.py:100 output: 'H' << 'e'
  DEBUG    root:machine.py:138 {TICK: 22, ADDR: 1, PC: 2, OUT: 101, ACC: 101} 
  DEBUG    root:machine.py:138 {TICK: 23, ADDR: 1, PC: 3, OUT: 101, ACC: 101} 
  DEBUG    root:machine.py:138 {TICK: 24, ADDR: 13, PC: 3, OUT: 1, ACC: 1} 
  DEBUG    root:machine.py:138 {TICK: 25, ADDR: 1, PC: 3, OUT: 101, ACC: 101} 
  DEBUG    root:machine.py:138 {TICK: 26, ADDR: 16, PC: 3, OUT: 0, ACC: -101} 
  DEBUG    root:machine.py:138 {TICK: 27, ADDR: 13, PC: 3, OUT: 1, ACC: -101} 
  DEBUG    root:machine.py:138 {TICK: 28, ADDR: 13, PC: 3, OUT: 2, ACC: -101} 
  DEBUG    root:machine.py:138 {TICK: 29, ADDR: 13, PC: 4, OUT: 2, ACC: -101} 
  DEBUG    root:machine.py:138 {TICK: 30, ADDR: 13, PC: 5, OUT: 2, ACC: -101} 
  DEBUG    root:machine.py:138 {TICK: 31, ADDR: 13, PC: 2, OUT: 2, ACC: -101} 
  DEBUG    root:machine.py:138 {TICK: 32, ADDR: 13, PC: 2, OUT: 2, ACC: 2} 
  DEBUG    root:machine.py:138 {TICK: 33, ADDR: 2, PC: 2, OUT: 108, ACC: 108} 
  DEBUG    root:machine.py:100 output: 'He' << 'l'
  DEBUG    root:machine.py:138 {TICK: 34, ADDR: 2, PC: 2, OUT: 108, ACC: 108} 
  DEBUG    root:machine.py:138 {TICK: 35, ADDR: 2, PC: 3, OUT: 108, ACC: 108} 
  DEBUG    root:machine.py:138 {TICK: 36, ADDR: 13, PC: 3, OUT: 2, ACC: 2} 
  DEBUG    root:machine.py:138 {TICK: 37, ADDR: 2, PC: 3, OUT: 108, ACC: 108} 
  DEBUG    root:machine.py:138 {TICK: 38, ADDR: 16, PC: 3, OUT: 0, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 39, ADDR: 13, PC: 3, OUT: 2, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 40, ADDR: 13, PC: 3, OUT: 3, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 41, ADDR: 13, PC: 4, OUT: 3, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 42, ADDR: 13, PC: 5, OUT: 3, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 43, ADDR: 13, PC: 2, OUT: 3, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 44, ADDR: 13, PC: 2, OUT: 3, ACC: 3} 
  DEBUG    root:machine.py:138 {TICK: 45, ADDR: 3, PC: 2, OUT: 108, ACC: 108} 
  DEBUG    root:machine.py:100 output: 'Hel' << 'l'
  DEBUG    root:machine.py:138 {TICK: 46, ADDR: 3, PC: 2, OUT: 108, ACC: 108} 
  DEBUG    root:machine.py:138 {TICK: 47, ADDR: 3, PC: 3, OUT: 108, ACC: 108} 
  DEBUG    root:machine.py:138 {TICK: 48, ADDR: 13, PC: 3, OUT: 3, ACC: 3} 
  DEBUG    root:machine.py:138 {TICK: 49, ADDR: 3, PC: 3, OUT: 108, ACC: 108} 
  DEBUG    root:machine.py:138 {TICK: 50, ADDR: 16, PC: 3, OUT: 0, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 51, ADDR: 13, PC: 3, OUT: 3, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 52, ADDR: 13, PC: 3, OUT: 4, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 53, ADDR: 13, PC: 4, OUT: 4, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 54, ADDR: 13, PC: 5, OUT: 4, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 55, ADDR: 13, PC: 2, OUT: 4, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 56, ADDR: 13, PC: 2, OUT: 4, ACC: 4} 
  DEBUG    root:machine.py:138 {TICK: 57, ADDR: 4, PC: 2, OUT: 111, ACC: 111} 
  DEBUG    root:machine.py:100 output: 'Hell' << 'o'
  DEBUG    root:machine.py:138 {TICK: 58, ADDR: 4, PC: 2, OUT: 111, ACC: 111} 
  DEBUG    root:machine.py:138 {TICK: 59, ADDR: 4, PC: 3, OUT: 111, ACC: 111} 
  DEBUG    root:machine.py:138 {TICK: 60, ADDR: 13, PC: 3, OUT: 4, ACC: 4} 
  DEBUG    root:machine.py:138 {TICK: 61, ADDR: 4, PC: 3, OUT: 111, ACC: 111} 
  DEBUG    root:machine.py:138 {TICK: 62, ADDR: 16, PC: 3, OUT: 0, ACC: -111} 
  DEBUG    root:machine.py:138 {TICK: 63, ADDR: 13, PC: 3, OUT: 4, ACC: -111} 
  DEBUG    root:machine.py:138 {TICK: 64, ADDR: 13, PC: 3, OUT: 5, ACC: -111} 
  DEBUG    root:machine.py:138 {TICK: 65, ADDR: 13, PC: 4, OUT: 5, ACC: -111} 
  DEBUG    root:machine.py:138 {TICK: 66, ADDR: 13, PC: 5, OUT: 5, ACC: -111} 
  DEBUG    root:machine.py:138 {TICK: 67, ADDR: 13, PC: 2, OUT: 5, ACC: -111} 
  DEBUG    root:machine.py:138 {TICK: 68, ADDR: 13, PC: 2, OUT: 5, ACC: 5} 
  DEBUG    root:machine.py:138 {TICK: 69, ADDR: 5, PC: 2, OUT: 32, ACC: 32} 
  DEBUG    root:machine.py:100 output: 'Hello' << ' '
  DEBUG    root:machine.py:138 {TICK: 70, ADDR: 5, PC: 2, OUT: 32, ACC: 32} 
  DEBUG    root:machine.py:138 {TICK: 71, ADDR: 5, PC: 3, OUT: 32, ACC: 32} 
  DEBUG    root:machine.py:138 {TICK: 72, ADDR: 13, PC: 3, OUT: 5, ACC: 5} 
  DEBUG    root:machine.py:138 {TICK: 73, ADDR: 5, PC: 3, OUT: 32, ACC: 32} 
  DEBUG    root:machine.py:138 {TICK: 74, ADDR: 16, PC: 3, OUT: 0, ACC: -32} 
  DEBUG    root:machine.py:138 {TICK: 75, ADDR: 13, PC: 3, OUT: 5, ACC: -32} 
  DEBUG    root:machine.py:138 {TICK: 76, ADDR: 13, PC: 3, OUT: 6, ACC: -32} 
  DEBUG    root:machine.py:138 {TICK: 77, ADDR: 13, PC: 4, OUT: 6, ACC: -32} 
  DEBUG    root:machine.py:138 {TICK: 78, ADDR: 13, PC: 5, OUT: 6, ACC: -32} 
  DEBUG    root:machine.py:138 {TICK: 79, ADDR: 13, PC: 2, OUT: 6, ACC: -32} 
  DEBUG    root:machine.py:138 {TICK: 80, ADDR: 13, PC: 2, OUT: 6, ACC: 6} 
  DEBUG    root:machine.py:138 {TICK: 81, ADDR: 6, PC: 2, OUT: 119, ACC: 119} 
  DEBUG    root:machine.py:100 output: 'Hello ' << 'w'
  DEBUG    root:machine.py:138 {TICK: 82, ADDR: 6, PC: 2, OUT: 119, ACC: 119} 
  DEBUG    root:machine.py:138 {TICK: 83, ADDR: 6, PC: 3, OUT: 119, ACC: 119} 
  DEBUG    root:machine.py:138 {TICK: 84, ADDR: 13, PC: 3, OUT: 6, ACC: 6} 
  DEBUG    root:machine.py:138 {TICK: 85, ADDR: 6, PC: 3, OUT: 119, ACC: 119} 
  DEBUG    root:machine.py:138 {TICK: 86, ADDR: 16, PC: 3, OUT: 0, ACC: -119} 
  DEBUG    root:machine.py:138 {TICK: 87, ADDR: 13, PC: 3, OUT: 6, ACC: -119} 
  DEBUG    root:machine.py:138 {TICK: 88, ADDR: 13, PC: 3, OUT: 7, ACC: -119} 
  DEBUG    root:machine.py:138 {TICK: 89, ADDR: 13, PC: 4, OUT: 7, ACC: -119} 
  DEBUG    root:machine.py:138 {TICK: 90, ADDR: 13, PC: 5, OUT: 7, ACC: -119} 
  DEBUG    root:machine.py:138 {TICK: 91, ADDR: 13, PC: 2, OUT: 7, ACC: -119} 
  DEBUG    root:machine.py:138 {TICK: 92, ADDR: 13, PC: 2, OUT: 7, ACC: 7} 
  DEBUG    root:machine.py:138 {TICK: 93, ADDR: 7, PC: 2, OUT: 111, ACC: 111} 
  DEBUG    root:machine.py:100 output: 'Hello w' << 'o'
  DEBUG    root:machine.py:138 {TICK: 94, ADDR: 7, PC: 2, OUT: 111, ACC: 111} 
  DEBUG    root:machine.py:138 {TICK: 95, ADDR: 7, PC: 3, OUT: 111, ACC: 111} 
  DEBUG    root:machine.py:138 {TICK: 96, ADDR: 13, PC: 3, OUT: 7, ACC: 7} 
  DEBUG    root:machine.py:138 {TICK: 97, ADDR: 7, PC: 3, OUT: 111, ACC: 111} 
  DEBUG    root:machine.py:138 {TICK: 98, ADDR: 16, PC: 3, OUT: 0, ACC: -111} 
  DEBUG    root:machine.py:138 {TICK: 99, ADDR: 13, PC: 3, OUT: 7, ACC: -111} 
  DEBUG    root:machine.py:138 {TICK: 100, ADDR: 13, PC: 3, OUT: 8, ACC: -111} 
  DEBUG    root:machine.py:138 {TICK: 101, ADDR: 13, PC: 4, OUT: 8, ACC: -111} 
  DEBUG    root:machine.py:138 {TICK: 102, ADDR: 13, PC: 5, OUT: 8, ACC: -111} 
  DEBUG    root:machine.py:138 {TICK: 103, ADDR: 13, PC: 2, OUT: 8, ACC: -111} 
  DEBUG    root:machine.py:138 {TICK: 104, ADDR: 13, PC: 2, OUT: 8, ACC: 8} 
  DEBUG    root:machine.py:138 {TICK: 105, ADDR: 8, PC: 2, OUT: 114, ACC: 114} 
  DEBUG    root:machine.py:100 output: 'Hello wo' << 'r'
  DEBUG    root:machine.py:138 {TICK: 106, ADDR: 8, PC: 2, OUT: 114, ACC: 114} 
  DEBUG    root:machine.py:138 {TICK: 107, ADDR: 8, PC: 3, OUT: 114, ACC: 114} 
  DEBUG    root:machine.py:138 {TICK: 108, ADDR: 13, PC: 3, OUT: 8, ACC: 8} 
  DEBUG    root:machine.py:138 {TICK: 109, ADDR: 8, PC: 3, OUT: 114, ACC: 114} 
  DEBUG    root:machine.py:138 {TICK: 110, ADDR: 16, PC: 3, OUT: 0, ACC: -114} 
  DEBUG    root:machine.py:138 {TICK: 111, ADDR: 13, PC: 3, OUT: 8, ACC: -114} 
  DEBUG    root:machine.py:138 {TICK: 112, ADDR: 13, PC: 3, OUT: 9, ACC: -114} 
  DEBUG    root:machine.py:138 {TICK: 113, ADDR: 13, PC: 4, OUT: 9, ACC: -114} 
  DEBUG    root:machine.py:138 {TICK: 114, ADDR: 13, PC: 5, OUT: 9, ACC: -114} 
  DEBUG    root:machine.py:138 {TICK: 115, ADDR: 13, PC: 2, OUT: 9, ACC: -114} 
  DEBUG    root:machine.py:138 {TICK: 116, ADDR: 13, PC: 2, OUT: 9, ACC: 9} 
  DEBUG    root:machine.py:138 {TICK: 117, ADDR: 9, PC: 2, OUT: 108, ACC: 108} 
  DEBUG    root:machine.py:100 output: 'Hello wor' << 'l'
  DEBUG    root:machine.py:138 {TICK: 118, ADDR: 9, PC: 2, OUT: 108, ACC: 108} 
  DEBUG    root:machine.py:138 {TICK: 119, ADDR: 9, PC: 3, OUT: 108, ACC: 108} 
  DEBUG    root:machine.py:138 {TICK: 120, ADDR: 13, PC: 3, OUT: 9, ACC: 9} 
  DEBUG    root:machine.py:138 {TICK: 121, ADDR: 9, PC: 3, OUT: 108, ACC: 108} 
  DEBUG    root:machine.py:138 {TICK: 122, ADDR: 16, PC: 3, OUT: 0, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 123, ADDR: 13, PC: 3, OUT: 9, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 124, ADDR: 13, PC: 3, OUT: 10, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 125, ADDR: 13, PC: 4, OUT: 10, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 126, ADDR: 13, PC: 5, OUT: 10, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 127, ADDR: 13, PC: 2, OUT: 10, ACC: -108} 
  DEBUG    root:machine.py:138 {TICK: 128, ADDR: 13, PC: 2, OUT: 10, ACC: 10} 
  DEBUG    root:machine.py:138 {TICK: 129, ADDR: 10, PC: 2, OUT: 100, ACC: 100} 
  DEBUG    root:machine.py:100 output: 'Hello worl' << 'd'
  DEBUG    root:machine.py:138 {TICK: 130, ADDR: 10, PC: 2, OUT: 100, ACC: 100} 
  DEBUG    root:machine.py:138 {TICK: 131, ADDR: 10, PC: 3, OUT: 100, ACC: 100} 
  DEBUG    root:machine.py:138 {TICK: 132, ADDR: 13, PC: 3, OUT: 10, ACC: 10} 
  DEBUG    root:machine.py:138 {TICK: 133, ADDR: 10, PC: 3, OUT: 100, ACC: 100} 
  DEBUG    root:machine.py:138 {TICK: 134, ADDR: 16, PC: 3, OUT: 0, ACC: -100} 
  DEBUG    root:machine.py:138 {TICK: 135, ADDR: 13, PC: 3, OUT: 10, ACC: -100} 
  DEBUG    root:machine.py:138 {TICK: 136, ADDR: 13, PC: 3, OUT: 11, ACC: -100} 
  DEBUG    root:machine.py:138 {TICK: 137, ADDR: 13, PC: 4, OUT: 11, ACC: -100} 
  DEBUG    root:machine.py:138 {TICK: 138, ADDR: 13, PC: 5, OUT: 11, ACC: -100} 
  DEBUG    root:machine.py:138 {TICK: 139, ADDR: 13, PC: 2, OUT: 11, ACC: -100} 
  DEBUG    root:machine.py:138 {TICK: 140, ADDR: 13, PC: 2, OUT: 11, ACC: 11} 
  DEBUG    root:machine.py:138 {TICK: 141, ADDR: 11, PC: 2, OUT: 33, ACC: 33} 
  DEBUG    root:machine.py:100 output: 'Hello world' << '!'
  DEBUG    root:machine.py:138 {TICK: 142, ADDR: 11, PC: 2, OUT: 33, ACC: 33} 
  DEBUG    root:machine.py:138 {TICK: 143, ADDR: 11, PC: 3, OUT: 33, ACC: 33} 
  DEBUG    root:machine.py:138 {TICK: 144, ADDR: 13, PC: 3, OUT: 11, ACC: 11} 
  DEBUG    root:machine.py:138 {TICK: 145, ADDR: 11, PC: 3, OUT: 33, ACC: 33} 
  DEBUG    root:machine.py:138 {TICK: 146, ADDR: 16, PC: 3, OUT: 0, ACC: -33} 
  DEBUG    root:machine.py:138 {TICK: 147, ADDR: 13, PC: 3, OUT: 11, ACC: -33} 
  DEBUG    root:machine.py:138 {TICK: 148, ADDR: 13, PC: 3, OUT: 12, ACC: -33} 
  DEBUG    root:machine.py:138 {TICK: 149, ADDR: 13, PC: 4, OUT: 12, ACC: -33} 
  DEBUG    root:machine.py:138 {TICK: 150, ADDR: 13, PC: 5, OUT: 12, ACC: -33} 
  DEBUG    root:machine.py:138 {TICK: 151, ADDR: 13, PC: 2, OUT: 12, ACC: -33} 
  DEBUG    root:machine.py:138 {TICK: 152, ADDR: 13, PC: 2, OUT: 12, ACC: 12} 
  DEBUG    root:machine.py:138 {TICK: 153, ADDR: 12, PC: 2, OUT: 0, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 154, ADDR: 12, PC: 2, OUT: 0, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 155, ADDR: 12, PC: 3, OUT: 0, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 156, ADDR: 13, PC: 3, OUT: 12, ACC: 12} 
  DEBUG    root:machine.py:138 {TICK: 157, ADDR: 12, PC: 3, OUT: 0, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 158, ADDR: 16, PC: 3, OUT: 0, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 159, ADDR: 13, PC: 3, OUT: 12, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 160, ADDR: 13, PC: 3, OUT: 13, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 161, ADDR: 13, PC: 4, OUT: 13, ACC: 0} 
  DEBUG    root:machine.py:138 {TICK: 162, ADDR: 13, PC: 6, OUT: 13, ACC: 0} 
  INFO     root:machine.py:291 output_buffer: 'Hello world!'
